### ISSUE TITLE
Unable to create remote dba user

### ISSUE THREAD

**Issue by littleprivate** [severalnines/s9s-tools#2]
Hi everybody,

If I try to create a remote dba user, I'll always get an error 256 like this:

s9s user --generate-key --create --cmon-user=username --controller="https://controller-IP:9501"
Connection to controller-IP closed.
SSH command exited with non-zero status: 256
ssh -tt -oUserKnownHostsFile=/dev/null -oStrictHostKeyChecking=no -oBatchMode=yes -oPasswordAuthentication=no -oConnectTimeout=30  'controller-IP' 'echo "{
    \"pubkey\": \"-----BEGIN RSA PUBLIC KEY-----\[......]n-----END RSA PUBLIC KEY-----\n\",
    \"user_name\": \"username\"
}" | sudo -n tee /var/lib/cmon/usermgmt.fifo 2>/dev/null >/dev/null'

Passwordless authentication from the remote client and the controller is enabled and working .... so I've no clue why the command fails :-(

**Comment by lpere** [comment]
Hi!
The command s9s tries to execute is 
  sudo -n tee /var/lib/cmon/usermgmt.fifo
It tries to write the JSON string into this named pipe. Maybe you could check that the pipe is actually there and writable.

As an alternative you could cut the command from the debug message, remove the "2>/dev/null >/dev/null" and try to run it manually. It might not make it work, but it could produce some error messages that help.

I think I will check the code and make the output available somehow, it is not very user friendly that the error messages are going all to /dev/null.

**Comment by littleprivate** [comment]
Hi!
If I execute the s9s command on the monitor server, eveything is working fine. But if I install s9s on my management server (another linux machine), I'll get this error. 
I understood the manual, that I'm able to install s9s on any machine in my network and be able to monitor the cluster remotely (especially 'cause passwordless ssh connection must be configured for remote access). On my management server is no /var/lib/cmon/usermgmt.fifo, because cmon isn't installed on this system. Cmon and s9s must be installed on the same system?

**Comment by lpere** [comment]
Hi!
The s9s program is a client that communicates with the cmon controller using https protocol. It does not communicate with any managed servers. The fifo is only used to bootstrap the system, to create the first cmon user (a user managed by the cmon controller).

If you already managed to connect to a the controller from an other computer I suggest you check the s9s configuration on the computer where the s9s worked. There should be a ~/.s9s/ directory on that computer, it should contain the configuration and the public/private key file that you used (assuming you used one and did not use username/password authentication). If you use the same schema you should be able to connect to the controller and that's you only need.

So to summarize:
1) You only need the fifo when you don't have username/password or username/keyfiles to authenticate and you can only use it to create new users.

2) The username/password and the username/keyfile also can be passed in the command line as options, but it is more practical to set these in the s9s configuration under ~/.s9s/.

3) If you check the ~/.s9s/ files you will see how the authentication towards the controller is set up. If you do the same on any new computer (same username, same pssword or same private key) the authentication should work.

4) The passwordless ssh only needed in one situation: when you have no cmon user account configured, you try to create a new account and the controller is a remote computer. Then s9s can reach the fifo only through an ssh command to create the account.

**Comment by kedazo** [comment]
Hey @littleprivate !

Can you please verify and make sure that SSH and sudo works to controller?
for example by running this?
```
ssh -tt 'controller-IP' 'sudo -n echo "Sudo is working" || echo "Not working."' 
```

Or you could re-run the command returned by s9s CLI in your first comment without 2>/dev/null as @lpere  suggested before..

If both SSH and sudo is working, and you have running enough new clustercontrol-controller (>= 1.4.2) with proper bind-addresses setup on your controller host then the user creation should work.

To set up the RPC to be available outside controller, you have to modify the controller's bind address
```
# you can specify certain IP's as comma separated list, by default cmon is listening only on 127.0.0.1
echo 'RPC_BIND_ADDRESSES="0.0.0.0"' > sudo tee /etc/default/cmon
sudo /etc/init.d/cmon restart
```

Best regards,
David

**Comment by kedazo** [comment]
Alternatively you could copy over from the working machine (monitor, where controller runs) the ~/.s9s directory to your management server (the other server), and adjust the hostname in ~/.s9s/s9s.conf controller= field... so SSH/sudo is not even needed in this case.

**Comment by littleprivate** [comment]
Hi David, 

I found a workaround for my problem ... the issue was, to use ansible for updating the cluster nodes and the monitor itself. I decided to execute the s9s commands for taking the nodes in maintenance on the monitor (by the delegate_to Ansible directive). This has the advantage, that there's no need to open the cmon management port on the monitor server. That works for me fine :-)
Your suggestion "ssh -tt 'controller-IP' 'sudo -n echo "Sudo is working" || echo "Not working."' " didn't work, I think it was caused of a missing password for become sudo. Anyway I'm no able to manage all cluster nodes with s9s via Ansible.

Bet regards, 
Andreas
